cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,883 @@
+
+        super().replace_node(node_id, new_node_id, variable_type=variable_type, meta=meta)
+
+    @_reset_attributes
+    def delete_node(self, identifier: NodeLike):
+        """
+        Delete a node from the graph.
+
+        See `cai_causal_graph.causal_graph.CausalGraph.delete_node` for more details.
+        """
+        super().delete_node(identifier)
+
+    @_reset_attributes
+    def delete_edge(self, source: NodeLike, destination: NodeLike):
+        """
+        Delete an edge from the graph.
+
+        See `cai_causal_graph.causal_graph.CausalGraph.delete_edge` for more details.
+        """
+        super().delete_edge(source, destination)
+
+    @_reset_attributes
+    def add_edge(
+        self,
+        /,
+        source: Optional[NodeLike] = None,
+        destination: Optional[NodeLike] = None,
+        *,
+        edge_type: EDGE_T = EDGE_T.DIRECTED_EDGE,
+        meta: Optional[dict] = None,
+        edge: Optional[Edge] = None,
+        **kwargs,
+    ) -> Edge:
+        """
+        Add an edge to the graph. See `cai_causal_graph.causal_graph.CausalGraph.add_edge` for more details.
+
+        In addition to the `CausalGraph.add_edge` method, this method also populates the metadata of the nodes with
+        the variable name and the time lag.
+        """
+        # if the source and destination time series nodes do not exist, create them
+        if source is not None and not self.node_exists(source):
+            source_meta = None
+            if isinstance(source, HasMetadata):
+                source_meta = source.get_metadata()
+                source = self.add_node(source, meta=source_meta)
+
+        if destination is not None and not self.node_exists(destination):
+            destination_meta = None
+            if isinstance(destination, HasMetadata):
+                destination_meta = destination.get_metadata()
+                destination = self.add_node(destination, meta=destination_meta)
+
+        edge = super().add_edge(source, destination, edge_type=edge_type, meta=meta, edge=edge, **kwargs)
+
+        return edge
+
+    @classmethod
+    def from_causal_graph(cls, causal_graph: CausalGraph) -> TimeSeriesCausalGraph:
+        """
+        Return a time series causal graph from a causal graph.
+
+        This is useful for converting a causal graph from a single time step into a time series causal graph.
+
+        :param causal_graph: The causal graph.
+        """
+
+        sepsets = deepcopy(causal_graph._sepsets)
+
+        # copy nodes and make them TimeSeriesNodes
+        ts_cg = cls()
+        for node in causal_graph.get_nodes():
+            meta = deepcopy(node.meta)
+            # get the variable name and lag from the node name
+            variable_name, lag = get_variable_name_and_lag(node.identifier)
+            node = TimeSeriesNode(variable_name=variable_name, time_lag=lag, meta=meta)
+            ts_cg.add_node(node)
```

Comment:
Why construct the node outside of the graph and not let the graph do it? For example `add_node` does not expect an instance of the node. We need to change the typedef on the add or change the logic here.